pkg/daemon/bgp/bgp.go

```diff
@@ -0,0 +1,353 @@
+/*
+ Copyright 2021 The Hybridnet Authors.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package bgp
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	daemonutils "github.com/alibaba/hybridnet/pkg/daemon/utils"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/alibaba/hybridnet/pkg/daemon/containernetwork"
+
+	"github.com/go-logr/logr"
+
+	api "github.com/osrg/gobgp/v3/api"
+	"github.com/osrg/gobgp/v3/pkg/server"
+)
+
+// Multiple bgp peers is not supported right now.
+
+type Manager struct {
+	localASN             uint32
+	peeringInterfaceName string
+
+	routerID string
+	// choose next hop address when advertise ipv4 address
+	routerV4Address net.IP
+	// choose next hop address when advertise ipv6 address
+	routerV6Address net.IP
+
+	bgpServer *server.BgpServer
+
+	logger logr.Logger
+
+	peerMap       map[string]*peerInfo
+	subnetMap     map[string]*net.IPNet
+	ipInstanceMap map[string]net.IP
+}
+
+func NewManager(peeringInterfaceName, grpcListenAddress string, logger logr.Logger) (*Manager, error) {
+	manager := &Manager{
+		// For using gobgp cmd to debug
+		bgpServer: server.NewBgpServer(
+			server.GrpcListenAddress(grpcListenAddress),
+			server.LoggerOption(&bgpLogger{logger: logger.WithName("gobgpd")}),
+		),
+
+		logger:               logger,
+		peeringInterfaceName: peeringInterfaceName,
+
+		peerMap:       map[string]*peerInfo{},
+		subnetMap:     map[string]*net.IPNet{},
+		ipInstanceMap: map[string]net.IP{},
+	}
+
+	peeringLink, err := netlink.LinkByName(peeringInterfaceName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get bgp peering link %v: %v", peeringInterfaceName, err)
+	}
+
+	existLinkAddress, err := containernetwork.ListAllAddress(peeringLink)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get link address for bgp peering interface %v: %v", peeringInterfaceName, err)
+	}
+
+	switch len(existLinkAddress) {
+	case 0:
+		return nil, fmt.Errorf("there is no valid address on bpg peering interface")
+	case 1:
+		manager.routerID = existLinkAddress[0].IP.String()
+		if existLinkAddress[0].IP.To4() == nil {
+			manager.routerV6Address = existLinkAddress[0].IP
+		} else {
+			manager.routerV4Address = existLinkAddress[0].IP
+		}
+	case 2:
+		if (existLinkAddress[0].IP.To4() != nil && existLinkAddress[1].IP.To4() == nil) ||
```

Comment:
XOR here?

Reply:
NO, I just want one ipv4 address and one ipv6 address here.

Reply:
Fixed